### PR TITLE
Add `average` to curve metrics

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

I personally needed a way to average pr-curves and roc-curves to make it easy to visualize the overall performance even for many classes. So I found a procedure from [sklearn](https://scikit-learn.org/stable/auto_examples/model_selection/plot_roc.html) on how to do it and implemented it in TM.

<details>
  <summary>Before submitting</summary>

- [ ] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
:books: Documentation preview :books:: https://torchmetrics--2084.org.readthedocs.build/en/2084/

### CODE REVIEW

File: src/torchmetrics/classification/auroc.py [hunk 1]
```diff
@@ -266,13 +266,15 @@ def __init__(
         )
         if validate_args:
             _multiclass_auroc_arg_validation(num_classes, average, thresholds, ignore_index)
-        self.average = average
+        self.average = average  # type: ignore[assignment]
```
Comment (Borda) [inline src/torchmetrics/classification/auroc.py:269]:
```suggestion
        self.average: Optional[str] = average
```

Comment (Borda) [reply src/torchmetrics/classification/auroc.py:269]: also fine with ignoring it if this does not work...

File: src/torchmetrics/classification/average_precision.py [hunk 2]
```diff
@@ -264,13 +264,15 @@ def __init__(
         )
         if validate_args:
             _multiclass_average_precision_arg_validation(num_classes, average, thresholds, ignore_index)
-        self.average = average
+        self.average = average  # type: ignore[assignment]
```
Comment (Borda) [inline src/torchmetrics/classification/average_precision.py:267]:
```suggestion
        self.average: Optional[str] = average
```

File: src/torchmetrics/utilities/compute.py [hunk 3]
```diff
@@ -131,3 +131,30 @@ def auc(x: Tensor, y: Tensor, reorder: bool = False) -> Tensor:
...
+    """One-dimensional linear interpolation for monotonically increasing sample points.
+
+    Returns the one-dimensional piecewise linear interpolant to a function with
+    given discrete data points :math:`(xp, fp)`, evaluated at :math:`x`.
+
+    Adjusted version of this https://github.com/pytorch/pytorch/issues/50334#issuecomment-1000917964
+
+    Args:
+        x: the :math:`x`-coordinates at which to evaluate the interpolated
+           values.
```
Comment (Borda) [inline src/torchmetrics/utilities/compute.py:146]:
```suggestion
        x: the :math:`x`-coordinates at which to evaluate the interpolated values.
```

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Pull request merged**